Editor: show buildhelp under objects

akupar opened the pull request:
This changes the editor to display buildhelp also when an object is placed on the field. The buildhelp icon is shown half-sized in these cases.

tothxa commented:
This is very useful.

Could you please make the overlays semi-transparent instead of smaller though? (like the Amazons buildhelp for treetop sentries)

I'm not sure if it would look good, but it would be helpful to show both current and maximum...

hessenfarmer commented:
Hi akupar,
welcome in the community and thanks for contributing.    

We are really glad about any contribution. However I noticed one small nit in this PR: 
the PR contains the changes of another PR which is unrelated to this one. this bares the risk of changes in the other PR being overwritten by this one. So please remove the changes from #5219 from this PR please.

akupar committed:
Show semitransparent icons in buildhelp when the field has an object

akupar commented:
> Could you please make the overlays semi-transparent instead of smaller though? (like the Amazons buildhelp for treetop sentries)

Changed it to show them semi-transparent instead. I didn't know there was a similar thing already as I haven't yet played with the Amazons. It should now look the same.

> I'm not sure if it would look good, but it would be helpful to show both current and maximum...

I'll look in to this.

akupar commented:
Thanks, hessenfarmer! It should be fixed now.

tothxa commented:
I experimented a little for showing both current and maximum:

Draw current over max:
![current_over_max](https://user-images.githubusercontent.com/80712255/151548033-46497945-cbea-4916-a2da-3c03006d56ac.png)

Draw max over current:
![max_over_current](https://user-images.githubusercontent.com/80712255/151548034-5bdfa455-43af-4c4b-90db-54a69a59a509.png)

Draw current over max, but reduce size if current == medium to improve visibility:
![proposed](https://user-images.githubusercontent.com/80712255/151548036-66ae8604-df60-48e4-af4a-4ae7cb297427.png)

I think the first 2 are both bad, but I quite like the third one. But please wait for other opinions before changing anything.

Here are the patches for the third variant:
[patch_against_pr.txt](https://github.com/widelands/widelands/files/7958680/patch_against_pr.txt)
[patch_against_master.txt](https://github.com/widelands/widelands/files/7958679/patch_against_master.txt)

BTW Thank you very much for this PR. This had been on my todo list for a long time, but it would have taken me ages to find the relevant pieces of code.

frankystone commented:
I think also the third one is a good solution. It may be hard to find a good reduced size, in the image it looks too small to me.

This should get an entry in the "Show/Hide" menu, e.g. "Show maximum buildings space" or so, with an explaining tool-tipp: "Shows which building size can be placed if all immovables are harvested." or so.

I can also imagine a similar solution for bug #1475 like: In a game show semitransparent portspace-icon if buildhelp is on, regardless if other immovables stand on this field.

tothxa commented:
> I think also the third one is a good solution. It may be hard to find a good reduced size, in the image it looks too small to me.

Here are 2 more scales. Previous one was 80%.

90%:
![medium_0 9](https://user-images.githubusercontent.com/80712255/151609409-1dab51b1-d456-463b-8a11-783a916b9ea3.png)

95% (I think this covers too much):
![medium_0 95](https://user-images.githubusercontent.com/80712255/151609414-c01c83ad-1f32-4ffd-82be-505664b61d6b.png)

> This should get an entry in the "Show/Hide" menu, e.g. "Show maximum buildings space" or so, with an explaining tool-tipp: "Shows which building size can be placed if all immovables are harvested." or so.

Well, I was hoping that if we can find a solution that looks good, then we wouldn't need such an option. OTOH if that's a separate option, then we don't have to show max and current at the same time.

> I can also imagine a similar solution for bug #1475 like: In a game show semitransparent portspace-icon if buildhelp is on, regardless if other immovables stand on this field.

Actually I wanted to propose doing this for games too. (maybe with a control to turn it off in a game or scenario, and a toggle for maps whether it is on by default for that map, for those who want to make it a challenge to find portspaces or big building spaces) (I'd do that PR myself of course)

frankystone commented:
+1 for 90%

> > This should get an entry in the "Show/Hide" menu, e.g. "Show maximum buildings space" or so, with an explaining tool-tipp: "Shows which building size can be placed if all immovables are harvested." or so.
> 
> Well, I was hoping that if we can find a solution that looks good, then we wouldn't need such an option. OTOH if that's a separate option, then we don't have to show max and current at the same time.

The main reason for adding a new menu entry is: A new option explains what is shown. If there is no such option a maps author get prompted with too much information, leaving him helpless what meaning the semitransparent icons may have. At least for new map authors, imho.

tothxa commented:
> The main reason for adding a new menu entry is: A new option explains what is shown. If there is no such option a maps author get prompted with too much information, leaving him helpless what meaning the semitransparent icons may have. At least for new map authors, imho.

Hmm, also when it's added in-game it will be better as a separate toggle, because showing max over player infrastructure is useful, but showing it all the time (I usually have buildhelp shown) would probably be annoying.

@akupar: Sorry, I didn't mean to hijack your pull request. Are you interested in doing some of this work? If not, I think we can accept this as it is now, and then I'll do the rest.

akupar commented:
> > The main reason for adding a new menu entry is: A new option explains what is shown. If there is no such option a maps author get prompted with too much information, leaving him helpless what meaning the semitransparent icons may have. At least for new map authors, imho.
> 
> Hmm, also when it's added in-game it will be better as a separate toggle, because showing max over player infrastructure is useful, but showing it all the time (I usually have buildhelp shown) would probably be annoying.
> 
> @akupar: Sorry, I didn't mean to hijack your pull request. Are you interested in doing some of this work? If not, I think we can accept this as it is now, and then I'll do the rest.

No problem :) I was actually just working on adding the menu item. I was thinking that space would bring up the current buildcaps and shift-space the maxcaps. Should the maxcaps, if they don't match the current caps, still be semi-transparent?

akupar commented:
I don't know if I should open a new issue, but seeing those pictures it occured to me that the hotspot of the build cap icons is not at the base of the building, but at the center of the image (except for the flags). Is this intentional?

frankystone commented:
> Hmm, also when it's added in-game it will be better as a separate toggle, because showing max over player infrastructure is useful, but showing it all the time (I usually have buildhelp shown) would probably be annoying.

The discussion here is related to the editor. Please don't mix comments for an in-game solution to keep the discussion clean :-)

tothxa commented:
> I don't know if I should open a new issue, but seeing those pictures it occured to me that the hotspot of the build cap icons is not at the base of the building, but at the center of the image (except for the flags). Is this intentional?

Seems to be intentional. If you look at the code where the images are loaded, the flags get a special case for determining the hotspot. You may open a new issue as under discussion for this.

tothxa commented:
> I was thinking that space would bring up the current buildcaps and shift-space the maxcaps.

:+1: 

> Should the maxcaps, if they don't match the current caps, still be semi-transparent?

I'm not sure. If it's a separate toggle, then probably it should be:
- buildhelp only: as current master
- max only: either show all of them with full opacity, or as your current, ie. show opaque where current == max, and semi-transparent where max != current (but don't show current)
- both max&current: where max!=current, semi-transparent max overlaid by current, but medium current reduced to 90%

akupar committed:
Add toggle maximum building spaces to display menu

Add toggle maximum building spaces to display menu + a key
shortcut (shift-space). If both maximum and actual building
spaces are shown, show actual sligtly smaller. Show maximum
semi-transparent when field has immovable. Also fix #4335.

akupar commented:
> Seems to be intentional. If you look at the code where the images are loaded, the flags get a special case for determining the hotspot. You may open a new issue as under discussion for this.

I tought it might look better when there are two icons if they were aligned to the bottom, but I don't think it matters now when they are separate toggles.

akupar commented:
>     * buildhelp only: as current master
> 
>     * max only: either show all of them with full opacity, or as your current, ie. show opaque where current == max, and semi-transparent where max != current (but don't show current)
> 
>     * both max&current: where max!=current, semi-transparent max overlaid by current, but medium current reduced to 90%

It should work like this now.

tothxa reviewed (commented):
I hope I didn't mess it up... Should've checked it out after all...

tothxa commented:
> The discussion here is related to the editor. Please don't mix comments for an in-game solution to keep the discussion clean :-)

I'm just trying to make it future-proof. :P

akupar committed:
Update src/editor/editorinteractive.cc

Co-authored-by: Tóth András <80712255+tothxa@users.noreply.github.com>

akupar committed:
Update src/editor/editorinteractive.cc

Co-authored-by: Tóth András <80712255+tothxa@users.noreply.github.com>

akupar committed:
unique enum value to fix build error

akupar committed:
Changed display flags to use the system defined in interactive_base.h

Co-authored-by: Tóth András <80712255+tothxa@users.noreply.github.com>

tothxa commented:
Sorry, but how do you make these commits? Something's not right with your diffs, and automatic formatting doesn't work on them. As for formatting, please use tabs for primary indentation.

tothxa commented:
> I actually realized that I had misunderstood what #4335 was about and changed it completely. Not sure if the editor-specific flags should go to editorinteractive.h or interactive_base.h, though.

I'd vote for interactive_base.h. Not just because you're continuing that, but also because all of these could be useful in games as well. (don't implement them in games yet though -- as @frankystone wrote, let's stick to the editor in this PR) Also please make the maxbuildspace flag 128. Both because then it comes directly next to buildhelp and because it's the one most likely to be added to games too.

But please wait for opinion from others as well on this.
@widelands/developers

frankystone commented:
From time to time you should merge the master branch into this one.

tothxa commented:
> all of these could be useful in games as well.

Correction: Show resources is definitely not needed in normal games :) (only in debug mode maybe) so please make it the last, maybe even leaving some room.

akupar committed:
Merge branch 'master' into editor-buildhelp-under-objects

akupar committed:
Removed commented out code that is not needed. Changed order of setting display flags

matthiakl commented:
> Something's not right with your diffs, and automatic formatting doesn't work on them.

@akupar, please configure your fork as explained in the [README](https://github.com/widelands/widelands#code) to allow the formatting bot to push commits to your branch.

akupar committed:
Changed order of display flags in case they are used in the game

Co-authored-by: Tóth András <80712255+tothxa@users.noreply.github.com>

akupar commented:
> From time to time you should merge the master branch into this one.

> @akupar, please configure your fork as explained in the [README](https://github.com/widelands/widelands#code) to allow the formatting bot to push commits to your branch.

Thanks, done!

> Sorry, but how do you make these commits? Something's not right with your diffs, and automatic formatting doesn't work on them. 

I think that was because I made that one change straight through the Github interface, and there others with git.

> As for formatting, please use tabs for primary indentation.

I'm been trying to configure this, but I haven't yet found the right setting.

tothxa commented:
> I think that was because I made that one change straight through the Github interface, and there others with git.

Please make sure that you merge or pull any changes made on github to yor local git repo before you push anything back to github. If you have to force-push, then something's wrong in your workflow. Please don't amend or rebase or do anything that rewrites the commit history in a public branch, eg. any branch for which you opened a pull request. Don't be afraid to add new commits, any PR will be merged as a single commit to master.

If you need to experiment, you can create a separate working branch in your local repo, and only check out your modified files into your local PR branch when you're satisfied with them for pushing to github.

akupar commented:
> > If you have to force-push, then something's wrong in your workflow. Please don't amend or rebase or do anything that rewrites the commit history in a public branch, eg. any branch for which you opened a pull request. Don't be afraid to add new commits, any PR will be merged as a single commit to master.

Ok, thanks. I thought it's bad to have lots of small commits.

> If you need to experiment, you can create a separate working branch in your local repo, and only check out your modified files into your local PR branch when you're satisfied with them for pushing to github.

I try to do this next time, thanks!

tothxa reviewed (changes requested):
Almost ready to go. See comments inline.

I created an icon for the show/hide menu based on the buildhelp icon if you don't mind: 
![toggle_maxbuild.png](https://user-images.githubusercontent.com/80712255/152190853-58158ae8-a05d-4f2b-95b2-f8740a31f218.png)

Looks like this in use:
![toggle_maxbuild_screenshot](https://user-images.githubusercontent.com/80712255/152191324-d17c7d68-ba28-4976-b238-6efa305cad53.png)

akupar committed:
Improve TRANSLATORS commento

Co-authored-by: Tóth András <80712255+tothxa@users.noreply.github.com>

akupar committed:
Add tooltip to show max buildhelp

Co-authored-by: Tóth András <80712255+tothxa@users.noreply.github.com>

akupar committed:
Merge branch 'master' of https://github.com/widelands/widelands into editor-buildhelp-under-objects

akupar committed:
Merge branch 'editor-buildhelp-under-objects' of github.com:akupar/widelands into editor-buildhelp-under-objects

akupar committed:
Moved display flags from editor to base

akupar commented:
> I created an icon for the show/hide menu based on the buildhelp icon if you don't mind: ![toggle_maxbuild.png](https://user-images.githubusercontent.com/80712255/152190853-58158ae8-a05d-4f2b-95b2-f8740a31f218.png)

Great!

I think there should be everything now.

tothxa committed:
add icon
remove double empty lines

tothxa committed:
fix formatting

akupar committed:
Merge pull request #1 from tothxa/maxbuild

add icon and apply autoformat fixes

akupar committed:
Moved Translators' comment directly above its translation

akupar committed:
fixed formatting

Co-authored-by: Tóth András <80712255+tothxa@users.noreply.github.com>

tothxa reviewed (approved):
Now it should be all done. Thank you.

akupar commented:
Thanks for all your kind help @tothxa and others. I learnt a lot in here!

tothxa committed:
fix typo

tothxa committed:
Merge branch 'master' into editor-buildhelp-under-objects

hessenfarmer commented:
> Thanks for all your kind help @tothxa and others. I learnt a lot in here!

Thanks for your contribution as well from my side. One thing left to ask: How do you want to be listed in the credits? (if at all) nickname or realname or both?

akupar commented:
> How do you want to be listed in the credits? (if at all) nickname or realname or both?

Nickname would be fine, thanks!

frankystone commented:
@akupar @tothxa : Anyone volunteer to fix #1475 in a similar manner? Please set yourself as assignee then :-)
